File: py/1909.remove-one-element-to-make-the-array-strictly-increasing.py
#
# @lc app=leetcode id=1909 lang=python3
#
# [1909] Remove One Element to Make the Array Strictly Increasing
#

# @lc code=start
class Solution:
    def canBeIncreasing(self, nums: List[int]) -> bool:
        bound = -1
        error = 0
        for i in range(len(nums)):
            if nums[i] <= bound:
                if error == 1:
                    return False
                error = 1
                if i < 2 or nums[i-2] < nums[i]:
                    bound = nums[i]
            else:
                bound = nums[i]
        return True
# @lc code=end

File: cpp/1909.remove-one-element-to-make-the-array-strictly-increasing.cpp
/*
 * @lc app=leetcode id=1909 lang=cpp
 *
 * [1909] Remove One Element to Make the Array Strictly Increasing
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool canBeIncreasing(vector<int> &nums)
    {
        int bound = -1;
        int prev_bound = -1;
        int error = 0;
        for (auto n : nums)
        {
            if (n <= bound)
            {
                if (++error == 2)
                {
                    return false;
                }
                if (prev_bound < n)
                {
                    bound = n;
                }
            }
            else
            {
                prev_bound = bound;
                bound = n;
            }
        }
        return true;
    }
};
// @lc code=end
